### ISSUE TITLE
Batchnorm handles empty batches incorrectly 

### ISSUE THREAD

**Issue by mkillinger** [deepmind/sonnet#125]
Context: I have sparse example data with sub-structure (think of: ` message data { repeated float sub_0 = 0;  repeated float sub_1 = 1; ...} `) where each sub-field is processed by a different branch of the network. The number of items in the sub-field varies (using reduction operations to obtain a fixed size vector `[batch_size, embedding_dim]`) - especially the number of items can be 0 for one sub-field in which case the output shape would be `[0, embedding_dim]`.

While all operations can correctly handle the case of having "size 0" in the batch axis, the operation that causes issues is the update of moving mean/variance (because tf.nn.moments that is used in `build_batch_stats` returns NaNs).

I there should be a guard around this, such that no updates with NaN-values are made when the inputs are 0-sized.
